rust/src/services/messages/mod.rs

```diff
@@ -0,0 +1,344 @@
+                .unwrap();
+            <Self as Service<Traced<PreProcessorRequest>>>::call(&mut svc, message.map(Into::into))
+                .await
+                .map_err(Into::into)
+                .unwrap()
+        }
+        .instrument(span)
+        .await
+    }
+
+    async fn call_state_machine(&self, message: Traced<MessageOwned>) -> StateMachineResponse {
+        let span = message.span().clone();
+        let mut svc = self.clone();
+        async move {
+            poll_fn(|cx| <Self as Service<Traced<StateMachineRequest>>>::poll_ready(&mut svc, cx))
+                .await
+                .map_err(Into::into)
+                .unwrap();
+            <Self as Service<Traced<StateMachineRequest>>>::call(&mut svc, message.map(Into::into))
+                .await
+                .map_err(Into::into)
+                .unwrap()
+        }
+        .instrument(span)
+        .await
+    }
+}
+
+/// A service that processes requests from the beginning to the
+/// end. The processing is divided in three phases:
+///
+/// 1. The raw request (which is just a vector of bytes represented an
+///    encrypted message) goes through the `MessageParser` service,
+///    which decrypt the message, validates it, and parses it
+///
+/// 2. The message is passed to the `PreProcessor`, which depending on
+///    the message type performs some additional checks. The
+///    `PreProcessor` may also discard the message
+///
+/// 3. Finally, the message is handled by the `StateMachine` service.
+#[derive(Clone)]
+pub struct PetMessageService<MessageParser, PreProcessor, StateMachine> {
+    message_parser: MessageParser,
+    pre_processor: PreProcessor,
+    state_machine: StateMachine,
+}
+
+impl<MP, PP, SM>
+    PetMessageService<
+        TracingService<MP, MessageParserRequest>,
+        TracingService<PP, PreProcessorRequest>,
+        TracingService<SM, StateMachineRequest>,
+    >
+where
+    MP: Service<Traced<MessageParserRequest>, Response = MessageParserResponse>,
+    PP: Service<Traced<PreProcessorRequest>, Response = PreProcessorResponse>,
+    SM: Service<Traced<StateMachineRequest>, Response = StateMachineResponse>,
+{
+    /// Instantiate a new [`MessagesService`] with the given sub-services
+    pub fn new(message_parser: MP, pre_processor: PP, state_machine: SM) -> Self {
+        Self {
+            message_parser: with_tracing(message_parser),
+            pre_processor: with_tracing(pre_processor),
+            state_machine: with_tracing(state_machine),
+        }
+    }
+}
+
+#[rustfmt::skip]
+impl<MP, PP, SM> Service<Traced<MessageParserRequest>> for PetMessageService<MP, PP, SM>
+where
+    MP:
+        Service<Traced<MessageParserRequest>, Response = MessageParserResponse> + Clone + 'static + Send,
+        <MP as Service<Traced<MessageParserRequest>>>::Future: 'static + Send,
+        <MP as Service<Traced<MessageParserRequest>>>::Error: Into<Box<dyn ::std::error::Error + 'static + Sync + Send>>,
```

Comment:
i agree that turning off the `rustfmt` for those kind of heavy trait bounds is reasonable. but the indention is off one level which makes it quite hard to read. same for the two other impls below.

```suggestion
    MP: Service<Traced<MessageParserRequest>, Response = MessageParserResponse> + Clone + 'static + Send,
    <MP as Service<Traced<MessageParserRequest>>>::Future: 'static + Send,
    <MP as Service<Traced<MessageParserRequest>>>::Error: Into<Box<dyn ::std::error::Error + 'static + Sync + Send>>,
```

Reply:
Ah ok. I thought grouping the bounds by type parameters made that more readable actually. I'll undo it if that's confusing.

Reply:
to me the layout reads that `MP` is a type with 3 trait bounds, but actually there are three trait bounds for three different types (although they are tightly coupled). not too much of a big deal though.

Reply:
I removed the `rustfmt::skip` attribute, so it should look more familiar now